### ISSUE TITLE
Too much detail from TTS when next action is >10 km, and for very closely spaced turns, mostly due to too very long dest tags

### ISSUE THREAD

**Issue by HolgerJeromin** [osmandapp/Osmand#2377]
When I enter a motorway the text to speech system told me exactly which exit, direction and road I have to take in 50 km. This is far too much information. 

My build in navigation device from my car says:
"Stay on this road for a long time"

That's all I need at this point. It will tell me more details 3 km before next action is required.

**Comment by vshcherb** [comment]
Agree, I think I've heard something like 58 kilometers 300 meters.

**Comment by sonora** [comment]
Just to clarify: I guess we are not talking about something like "58 kilometers 300 meters" (which should not happen, at least not in the major tts languages I watch?? We always round to full km for >10), but we are talking about something like Osmand saying "Continue for 58 km on A3 to Rome, Flaminia Ave" (where the exact details of the exit description like "Rome, Flaminia Ave"  depend on its specification in the map and can in deed be even more lengthy in some cases)?

Question is how exactly to shorten this, are you saying we should unconditionally suppress the exit (or intersection) info entirely for distances >10km?

Trouble is: The info can be as precise and useful as saying "Rome, Flaminia Ave", which is a good exit specification, or could be as superfluous as saying "Rome, Flaminia Ave South, Flaminia Ave North, Olympic Stadium, Olympic Village, Flaminia train station", which is a lot more than you will want to know 10 km ahead ...

**Comment by vshcherb** [comment]
I think I will need to test again to find exact issue.

**Comment by sonora** [comment]
PS: The related bigger issue, which has bugged me for quite a while, but I find no good solution, is this:

Not for the long range case >10km, where you hear this lengthy prompt only once, but for the extreme short range case, like in a big "spaghetti-bowl" type motorway intersection, where lanes may split twice within 300m or so: You hear all these lengthy exit descriptions 3-4 times or so in almost overlapping voice prompts (turn, turn now, turn, turn now), all of which are necessary, but too long due to the long exit descriptions!

But not sure how to shorten this reliably, without deteriorating all other situations where it works good ...

**Comment by HolgerJeromin** [comment]
> we are talking about something like Osmand saying "Continue for 58 km on A3 to Rome, Flaminia Ave" (where the exact details of the exit description like "Rome, Flaminia Ave" depend on its specification in the map and can in deed be even more lengthy in some cases)?

Exactly this was my point. I think I can live with dropping all specific information.

**Comment by sonora** [comment]
I am really hesitant to simply drop any exit-info entirely for the long range prompt. Where the prompt is short and precise, like saying "Continue on A1 for 12 miles to Flaminia Avenue", I think it is really useful, because it is the only way (without many screen taps) to know how OsmAnd intends to route you, and you may have thoughts like "Yes, Flaminia Ave sounds like a great idea", or "hmm, Flaminia this time of day could mean a lot of traffic, I think I will leave the highway one exit before that and let OsmAnd re-route."

I will continue to look for a way to solve the underlying issue of maybe cutting too long exit descriptions to the essential, which would also solve the issue of lengthy back-to-back prompts in some motorway intersections in the very short-range, which I find much more annoying.

To demonstrate what I am talking about, I have looked for a test case, and think, this is one: Route from Lat=49.40832, Lon=11.32691 to Lat=49.40857, Lon=11.20285, taking 2 consecutive rights (where the lanes split each time, like on many German motorway intersections) and listen to the voice prompts, always containing the lengthy "...towards Berlin, Regensburg, etc".  the prompts are 100% correct, but at 100km/h the continuous talking is just very "crowded".

**Comment by raindropsfromsky** [comment]
IIRC this instruction was in the "Follow the road for x km" format.
That information relaxes you. Now you will stop wondering at each junction whether to take a turn.

On the other hand, the **next maneuver** announcement should not come more than 2-3 minutes before the actual event. Thus even the 10 km is not a good threshold, as I regularly face the voice instruction "turn left after 9 km"). In Bangalore, the average speeds are so low that this turn comes after a good 40 minutes! Why should I care that I have to turn left after 40 minutes??

**Comment by sonora** [comment]
Very preliminary implementation notes:

I guess the solution could be that we introduce a cut_part_street_shortened( ), which would essentially ignore "Dest", just concat "Ref" and (perhaps) "Name", and this shortened string could then be used instead of cut_part_street( ) for
- The long-distance "Continue" prompt
- In the short-succession case for the final "turn" prompt. We need to somehow find a way to discover the condition when the end of the previous prompt was only very shortly ago, i.e. when announcements are "very crowded"

Still hoping I find a better way to discover "crowded" announcement situations...

**Comment by raindropsfromsky** [comment]
I would like to volunteer to check this out, by simulating a journey through turns in quick succession. 

I could have done this using the old design, where the **OsmAnd Development ** plugin provided a "simulate" button on screen. But in the new design, this facility is buried deep. To launch it once, I have to use as many as 13 operations, and still the result is not satisfactory!

1. Tap on **Menu**
2. Tap on **Settings**
3. Tap on **Plugins**
4. Swipe down to the bottom of plugin list
5. Tap on **OsmAnd development**.
6. Tap **Settings**
7. Swipe down by one screen
8. Tap the **Simulate your position** option.
9. Tap **OK**.
10. Tap the **Back** button three times to return to map.
by this time, the simulation has already started!

If the plugin has an option to add the **Simulate** button on the screen, it would be great!

**Comment by sonora** [comment]
Narayan, I think the Dashboard has more readily accessible "Simulate" button if the Development Plugin is enabled.

After some testing I come to the following conclusion: In the 2 prompts
 - "Continue ..."  (which is the long-distance case) and
 - "Turn ..." (which is the very last, immediate distance turn prompt, without "after x meters" or "in x meters")
we should generally ignore all Destination information.
So these 2 prompts would generally be shortened to sound (maximum) something like
 - "Continue for 15 km to A3", or
 - "Turn right onto B4 Main Street", respectively.

All other prompts would remain at full length. Going through our code history, I believe the whole issue did start in the first place because we introduced the extensive "Destination" prompting some time back.

**Comment by Barrocas** [comment]
That depends on country and Destination info available on Osm.

I do like the announcement to Destination as it is, usually it causes me no problem.

Here in Portugal, the Destination is much more usefull as an indication prompt, since it's what is written on the road signs.

It's easier for the driver to hear the Osmand prompt, and read the according road sign to take the exit/turn.

The street name is mostly hard to find to a driver, usually it's a very small sign attached to a building on that street.

Maybe shorten the Destination announcement?
So if a Destination has some 5 destination, read only the first?
That still wouldn't such a complete option as we have now, but it would shorten things out and yet be useful.

In my opinion, using only the street name and not Destination (when available) would mostly result in way longer promps, at least around here.

**Comment by sonora** [comment]
Barrocas, I am not talking about generally changing anything about the destination announcements as we have them, I fully concur with you.

I am only talking, for the sake of more brevity, about dropping it from 
- the very long range "Continue..." prompt, as was requested in this original issue
- the very last "Turn (now)" prompt, where it is usually a repetition or even repeated repetition already, and in many cases so shortly before the actual turn you have no time to listen to it yet again and then still react to it, often lengthy dest info in the very last prompt lasts until you are already way in the turn anyway.

I have "tested" for the last few days how it would be if we dropped the Dest info from these 2 prompts, and my conclusion is it could greatly improve things. But please offer your opinions! To test, please simply imagine "imagine" the destination portion being dropped from these 2 prompts (only) during actual navigations.

**Comment by sonora** [comment]
PS: Re "speak only first destination": Please "mind-test" this also in as many situations as you can. Need to decide if it produces useful results, and also if to do it only in the 2 prompts I say we should maybe modify, or in  ALL prompts?

**Comment by Barrocas** [comment]
Sonora, I share your opinion, I have no problem whatsoever with a longer announcement to a long range destination, as you stated, it may even be better for deciding if the routing is suitable (or not).
to me it's not even an issue, but I understand if someone doesn't like it.

The very short range consecutive indications may be a bit more complicated, (and could be considered another case, although related). These could use some shorter indications if the  'destination' tag is too long. It would be a hard call between short and fast (but incomplete) indications, or longer and more complete indications...

**Comment by sonora** [comment]
Victor, I think I would like to give this a try as discussed in this thread.

I think the way to do it is in VoiceRouter.java, expand getSpeakableStreetName(currentSegment, next) to something like getSpeakableStreetName(currentSegment, next, boolean=fullDest), and then for fullDest=false either completely nullify "getSpeakablePointName(i.getDestinationName())", or maybe terminate it after the first semicolon or comma.

Would you say this is the best place and way to code it, or do you have any other idea? It appears that doing this in our ttsconfig.p would make the code there very convoluted ....

Thx - Hardy

**Comment by sonora** [comment]
I think I fixed this pretty well, both issues raised here. Please test and see if you find any disadvantages or if we need fine-tuning. For now I believe that leaving out the Dest info for the very long-haul and the very last "turn (now)" prompts is the way to go.

(Our spare option would be to cut the Dest info after its first separator sign, but that may have undesired effects, and may still be too lengthy for the very last prompt if accompanied by a lot of ref info as well.)

**Comment by Barrocas** [comment]
Sorry, but I don't like it.
Correct me if I'm wrong, but this is what I think:

1 - Long range Destination instructions - As said before, I didn't consider
this to be an issue at all.
It was a clear instruction to what destination exit was to be taken, and
the driver could then mentally associate the announced 'Destination'
(easier to understand than 'Ref' or 'name') with the route he intends to
follow , and if it sounded incorrect, he could change the route, it if
necessary.
Since it was a long range to drive, there was no problem whatsoever in
having a long and detailed instruction. I consider it was valuable
information.

2 - very short range - I also considered this important, because it was a
last resource indication if we were following the right way or not.
Sometimes I receive a phone call while driving, and cannot see the route
(phone changes screen to receive the call), nor clearly hear all the
previous indications provided by TTS (there's someone speaking in a call),
so a last turn 'Destination' warning was nice.
Sometimes the maps are not made in a way that's best for Osmand, and the
turn indications (left, right, etc) may come kind of strange, the
'Destination' instruction is clearer to follow for the driver.
I believe we also lost the roundabout exits 'Destination' warnings with the
changes.
And if the next turn is close, then the 'Destination' instruction is not
even spoken at all.

So to cut the chattiness, we lost lots of routing information.
I preffered the chatiness.



-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by HolgerJeromin** [comment]
My usecase for opening this issue was:
OsmAnd told me 40 km ahead something like (it was a few weeks ago and german):
Leave the motorway in 40 km on Motorway interchange "Aachener Kreuz" to direction "Europaplatz, Rothe Erde, Würseln, Antwerpen, Eindhoven, Heerlen, Aachen"

after listening this long listing i have forgotten the start of the sentence already.

**Comment by sonora** [comment]
Sorry for the long delay in answering. I have tested what we have now for some 5000 km in recent weeks, and in many countries, and have to say it is the best we have ever had in OsmAnd. Yes, it is true, on the odd occasion that you missed pre-announcements the final "Turn right onto Motorway 6" may be less informative than hearing the full "Turn right onto Motorway 6 towards Rome, Naples, Salento, Reggio Calabria". But then I find that in many turn cases, both in cities and on motorway intersections, there is little time for the final announcement, so the long versions often extended past when the turn actually had to be made anyway (plus you needed time to read signs again).

As I said, I test OsmAnd a lot, have done so for 5 years now, and you can always find this and that situation where something else would be better or worse. So what we cannot automate or deterministically foresee needs to be a compromise over the average cases. And I very strongly feel that on the vast majority of occasions what we have now is a clear improvement over what we ever had before ...  :-)

**Comment by Barrocas** [comment]
I agree it is less chattery, and I understand the point.

But I'm missing the destination spoken in at least 2 situations, where
it was much clearer to hear the destination to follow:

- at roundabouts
- after some motorway tolls, when roads split shortly after the toll.

we used to have the clear indication to wich destination to follow,
but now, since both usually are close exits, it only reads the ref,
wich in the motorway case, most times is the same for any of the
available roads (same highway, but diferent destinations).

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
Barrocas, it is true. But particularly for the roundabout case we had another explicit request somewhere to drop all additional info from the very last exit prompt, because in many roundabouts the time to react to that prompt is too short anyway and people requested a prompt as swift as possible.

Frankly, I do not know how to fulfill both requests at the same time. We could probably try trimming any destination info to the first occurrence of ";", but this would again result in "Turn right onto Highway ABC towards XYZ", so is likely half way back to where we came from.  If you have any good idea, please let me know.

Best - Hardy

**Comment by Barrocas** [comment]
I never had issues with the destination tags while driving.

Sure I understand other users wanting less chattering while driving,
especially if the 'destination' is very extensive as reported in other
cases, but it was really usefull, at least were I live, and also for
visiting drivers that are not familiar with the local roads.

As explained before, my main reason is that the 'destination' tag is
what is usually  written in the roadside indication boards, and it's
imediatly visible and recognizable while driving.
On the other hand, most of the times, the road 'name' or 'ref' is not
easily visible, or not even visible at all, so it's not near as good
as a driving indication.

In a roundabout, if it only says 'take the 3rd exit', that leads to
other issues like when there are incorrectly mapped areas, or new
unmapped exits, or even if I'm really bad at math and counting while
driving  :)

So in the 2 cases, of roundabouts or short range turns, the
'destination' info is tottally lost, even if it's correctly mapped.
Being a mapper, I sometimes even wonder if the mapping is correct or
not, since I was used to hearing 'destination' announcements that now
are completely absent.

If possible, I would prefer the option to the indication speaking at
least the first (Main) 'destination' only, on that short range (or
last turn) announcement, but I don't know if that's possible, or even
if other users would agree on this.

I do understand that the long indications may be a problem, but only
if there is the need for another indication/exit/turn shortly
afterwards. If not, I prefered the full indications like we had
before. Just my 2cents...

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
@vshcherb : Victor,I would welcome your opinion on this one. Looks like the current options are

(A) leave the fix as we have it now, with the disadvantages Barrocas explains

(B) Introduce some sort of setting if you want the Destination being read in the very last Turn/Exit prompts. This is not a user-specific topic, but rather a situation-specific issue, so we should try to find a way to resolve this situation-dependent, which I cannot see, except

(C) Re-introduce the destination, but shorten it to read only the first location, which still makes the prompts rather long again in many languages.

(D) I wish we could do the following: Find a way to read the destination in the last turn prompt (only) if it has not previously been read in a pre-announcement, maybe I will investigate this when I have time. Not sure it would well solve the motorway-cloverleaf issue, as there lanes fork several times so each turn look may look like a new turn without pre-announcement.

Best, Hardy

**Comment by vshcherb** [comment]
I'm might be wrong but I think destination should be pronounced if the street names are turned on, I personally disable street names because would like to hear sharp. Though if people are selecting speak street names,  I think we should speak destinations. Though I don't think we should speak both it might be another option and possibly in accessibility "speak everything possible"

**Comment by sonora** [comment]
To clarify:

As an example, here is a typical sequence approaching a motorway intersection where you take 2 consecutive right forks (really only lane changes) to change onto the intersecting motorway ("Highway 1"), first right fork (turn A) takes you on the exit ramp to change towards Highway 1, second right fork (turn B) takes you on the lane then entering Highway 1 in the correct direction (towards City 3 and City 4).

Prompts are as follows:
- (A0) "Continue for 12 kilometers to Highway 1"

- (A1) "After about 1 kilometer turn slightly right onto Highway 1 towards City 1, City 2, City 3, City 4."

- (A2) "In 500 meters turn slightly right onto Highway 1 towards City 1, City 2, City 3, City 4, then keep right."

- (A3) "Turn slightly right onto Highway 1 **(towards City 1, City 2, City 3, City 4.)**, then in 150 meters turn slightly right."

- (B3) "Turn right onto Highway 1 **(towards City 3, City 4.)**"

As a response to numerous issues raised, the portion between the parentheses **()** is NOT played in our current nightly builds, to avoid the possibly 3rd and 4th repetition of lengthy city lists, which annoyed many people including me. 

@Barrocas , would this scenario solve your issue: My idea is I could re-introduce to play the **()** portion under the precondition that there was no previous "Turn in" statement with a destination announcement for the very same turn (to avoid repetition). So this would re-introduce it in B3 above, but not in A3.

Equivalent rule for roundabout ("Take the n-th exit ....") scenario.

**Comment by Barrocas** [comment]
Hi,
Yes, your proposal seems better than what we have now.

So if I understood it correctly, you propose that in turn A3 we need a
short indication because there will be another turn shortly after (not
convenient to have a very long announcement), but the next turn B3 has the
full indication as we had before the changes.

The way I thought about it, is that the announcements could be in full, but
only if there won't be another next turn shortly after (or another
needed announcement).

So we kind of agree on this :)


-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
Ok, it is implemented, please give it a try.

The way it is now, it is ensured that you will hear the full destination announcement at least once or possibly twice, but never 3 or 4 times, as we had it before. (Although you may not hear it again on the very last turn prompt, if you already heard it once or twice before).Too many repetitions was the underlying issue many of our users had raised.

The way you suggest to make it dependent on _subsequent_ turns would not per se solve the issue of many repetitions, and as a rule hearing it on the last turn prompt only is way too late in many situations, I think. So I suggest the latest  nightly is probably the best way we could find.

**Comment by sonora** [comment]
PS: Please put some focus on testing the mechanism for roundabouts. I am very satisfied with how we have it now for "Turns", but find no sufficient assortment of roundabouts (we need to test all sizes, they come from mini to very large) with destinations mapped.

Current mechanism for roundabouts is the same as stated above for turns:
Destination is announced in the "After..." prompt (equivalent of A1) and the "In.." prompt (equivalent of A2), but for A3 ("Take the nth exit onto...") only if not already previously announced once or twice already.

**Comment by Barrocas** [comment]
Sure, I can test it this next week, and see how it's working.

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
@Barrocas, Thank you, much appreciated! I have re-read this long thread once more to make sure we finally captured everything, and have made 2 more changes, so please use the May 14 or later nightly for your tests:
-  No suppression of the destination in the last roundabout prompt,("Take the n-th exit"), because --other than for turns-- this is the worst case second time it will sound. (Please double-check if this is useful and not too chatty).
- I have added an exception to repeat the destination to the final turn prompt again if it is immediately after a toll booth, to honor your respective comment above.

Hope this works now, and it is not too chatty.

**Comment by Barrocas** [comment]
Thanks!

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by Barrocas** [comment]
It appears to be working as expected.

Roundabouts are ok, motorway exits after toll also ok, some even warn
before the toll and after the toll, just before we have to make a
decision, so that's fine.

I only found one situation where it could be better, and that was here:
http://www.openstreetmap.org/way/379672276

in this case, there are 2 possible motorway exits in a close distance (150m).
The firts exit is announced ok, but the second, as it is in short
range, is only announced when reaching the exit.

Before the changes were made, it would announce something like 'follow
the road for 150m to Meadela' for going straight on, and then announce
'turn right at A27 to Ponte de Lima'.
This way, it was clear for the driver what road to follow, otherwise
it's quite easy to just take the first exit in a mistake.

Now it only announces to 'follow the road for 150m' (no destination
announced), and then it announces only on the last turn to exit 'turn
right at A27 to Ponte de Lima'.

I believe that it would be better to either announce it as before
(destination to go straight, and then destination for the turn exit),
but since it hasn't got much time to speak the turn destination before
the exit, it would be better to just say something like:
'follow the road for 150m to Meadela and then turn right at A27 to
Ponte de Lima'.

So it could just say 'turn right at A27' on the turn exit, as in all
other motorway last turn announcements.


-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by Barrocas** [comment]
Or even simpler, ''follow the road for 150m and then turn right at A27 to
Ponte de Lima'

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
Ok, once again, thank you so much for testing! Looks like we have fine-tuned it pretty well now!

The situation you report http://www.openstreetmap.org/way/379672276, coming from the north and turning towards the east, is interesting, because it is actually affected by Issue #2571. I have added the destination announcement back to he "Continue" announcement, no problem there. So please use a May 21 or later build for further testing.

But perfectly, the situation should be like any other motorway exit, where the turn is properly pre-announced with full details (twice, if you come with a long approach), then announced (abbreviated). But this logic seems interrupted here by what I elsewhere call a "ghost turn", which creates a segment of saying "go straight" in our routing.

In my simulation I even seem to hear a surplus bogus "In 150 m turn right" (no more details) before the end of the straight segment and the actual right turn (which is then announced with full details, as the pre-announcement is missing),. Maybe you can state if you have the same finding,

I will  think about what can be done about it, maybe we will actually have to do something about Issue #2571.

**Comment by Barrocas** [comment]
Yes, it happens as you describe (the bogus turn).

No problem, I pass in that place almost everyday (except weekends),
and I usually take the first (A28 Porto) or the second exit (A27 Ponte
de Lima, the ghost turn), so it will be an easy task to test it :)

-- 
Cumprimentos,
D.er Hugo Barrocas

**Comment by sonora** [comment]
Ok, I debugged the "bogus turn", it is actually the lane selection "go straight" for the first exit, the one you are _not_ supposed to take when going east.

So there are 2 options now: We can either mute that bogus voice prompt, or what I have done now: I have converted it into a "meaningful" announcement saying "Go straight ahead, then in 150m turn right."  (This is the logical equivalent of what we do in such situations if there was no "straight", but say a right turn: We would announce "Turn right, then in 150m turn right.")

So I guess we could leave it as we have it now in May 22+ builds. It may sound rather crowded in this particular situation, that comes from the long Destination tag o the previous prompt "Viana de Castelo, Meadela, A27." But I suppose the additional "go straight" prompt after it may not even be so unwelcome for the driver, so Issue #2571 may even be a "feature" here, not a "bug".  :-)

If the situation sounds unacceptably crowded, please let me know, then we can try with entirely muting the former bogus prompt.

**Comment by sonora** [comment]
Ok, I am very confident we got this consistent in our code and it is correctly fine-tuned now.

Remaining chattiness from hearing occasional (and arguably obsolete) "Go straight ahead ..." prompts comes from Issue #2571.

For any other voice-navigation related issues, should anyone find something, I suggest opening a new issue.

**Comment by Lee-Carre** [comment]
@raindropsfromsky

> I could have done this using the old design, where the *OsmAnd Development * plugin provided a "simulate" button on screen. But in the new design, this facility is buried deep. To launch it once, I have to use as many as 13 operations, and still the result is not satisfactory!
> 
> 1. Tap on **Menu**
> 2. Tap on **Settings**
> 3. Tap on **Plugins**
> 4. Swipe down to the bottom of plugin list
> 5. Tap on **OsmAnd development**.
> 6. Tap **Settings**
> 7. Swipe down by one screen
> 8. Tap the **Simulate your position** option.
> 9. Tap **OK**.
> 10. Tap the **Back** button three times to return to map.
by this time, the simulation has already started!
> 
> If the plugin has an option to add the **Simulate** button on the screen, it would be great!

You'll be pleased to know that #7778 improves this.
